### PR TITLE
Materialize: Add additional comparison operators in Materialize and fix bug where they not applied for sharded keyspaces

### PR DESCRIPTION
**Pull request by rohit-nayak-ps**
## Description

This PR fixes a bug where comparison filters were being ignored in Materialize workflows if the source keyspace was sharded.

In addition, it adds support for additional operators (`>, >=, <, <=, <>`) . The filter expression can chain multiple comparison operations using `and` only.

Examples:
```
select * from t1 where id = 1
select * from t1 where id <> 1
select * from t1 where val > 'abc'
select * from t1 where id >= 1
select * from t1 where id < 2 and val <= 'xyz'
```
Signed-off-by: Rohit Nayak <rohit@planetscale.com>
## Checklist
- [X] Tests were added or are not required
- [ ] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vttablet/tabletserver/vstreamer/planbuilder.go [hunk 1]
```diff
@@ -111,6 +121,42 @@ func (plan *Plan) fields() []*querypb.Field {
...
+// compare returns true after applying the comparison specified in the Filter to the actual data in the column
+func compare(comparison Opcode, columnValue, filterValue sqltypes.Value) (bool, error) {
+	result, err := evalengine.NullsafeCompare(columnValue, filterValue)
+	if err != nil {
+		return false, err
+	}
+	if result == 0 && (comparison == Equal || comparison == LessThanEqual || comparison == GreaterThanEqual) ||
+		result < 0 && (comparison == LessThan || comparison == LessThanEqual || comparison == NotEqual) ||
+		result > 0 && (comparison == GreaterThan || comparison == GreaterThanEqual || comparison == NotEqual) {
+		return true, nil
```
Comment (shlomi-noach) [inline go/vt/vttablet/tabletserver/vstreamer/planbuilder.go:155]: Personal preference: break this `if` into 3 `if` statements (`== 0`, `< 0`, `> 0`), each of which ends with `return true, nil`.
Alternatively, a `switch` statement with 3 cases. 
There's just too much `&&` and `||` going on.

**Review by shlomi-noach — COMMENTED** [review]
I'm not sure what in this PR adds support for more Opcodes other than the parser part. I'm guessing they are automatically handled later on by the same existing code?

Otherwise I confess I don't understand the code nor the changes :/

File: go/vt/vttablet/tabletserver/vstreamer/planbuilder.go (hunk 1 above)
Comment (systay) [reply go/vt/vttablet/tabletserver/vstreamer/planbuilder.go:155]: another alternative is to use bitmasks for the comparisons, so you could write something like:

```
if result == 0  {
  return comparison =| 1 << Equal, nil
}
if result < 0  {
  return comparison =| 1 << LessThan, nil
}
if result > 0  {
  return comparison =| 1 << GreaterThan, nil, nil
}
```

Comment (shlomi-noach) [reply go/vt/vttablet/tabletserver/vstreamer/planbuilder.go:155]: I find this syntax confusing. Didn't ever see `=|` used before in `golang`

Comment (systay) [reply go/vt/vttablet/tabletserver/vstreamer/planbuilder.go:155]: how are you used to checking bitmasks?

Comment (shlomi-noach) [reply go/vt/vttablet/tabletserver/vstreamer/planbuilder.go:155]: C-style: `if my_var | 0x0040 { do_something }`

Comment (systay) [reply go/vt/vttablet/tabletserver/vstreamer/planbuilder.go:155]: sorry for the confusing comment. I was thinking of something like:

```golang
	const (
		NotEqual = 0
		Equal = 1 << iota
		LessThan
		GreaterThan
	)
	var cmp int
	var x sqlparser.ComparisonExpr
	switch x.Operator {
	case sqlparser.EqualOp:
		cmp = Equal
	case sqlparser.LessThanOp:
		cmp = LessThan
	case sqlparser.GreaterThanOp:
		cmp = GreaterThan
	case sqlparser.LessEqualOp:
		cmp = LessThan | Equal
	case sqlparser.GreaterEqualOp:
		cmp = GreaterThan | Equal
	case sqlparser.NotEqualOp:
		cmp = NotEqual
	}
	var result int
	if cmp == NotEqual {
		return result != 0, nil
	}
	if result == 0 {
		return cmp&Equal != 0
	}
	if result < 1 {
		return cmp&LessThan != 0
	}
	if result > 0 {
		return cmp&GreaterThan != 0
	}
```

Comment (rohit-nayak-ps) [reply go/vt/vttablet/tabletserver/vstreamer/planbuilder.go]: Simplifed the checks to a switch-based one. @shlomi-noach, `newVStreamer()` in vstreamer.go has an explanatory comment on `Filters`.

**Review by shlomi-noach — APPROVED** [review]

File: go/vt/vttablet/tabletserver/vstreamer/planbuilder.go [hunk 2]
```diff
@@ -111,6 +121,67 @@ func (plan *Plan) fields() []*querypb.Field {
...
+	default:
+		return -1, fmt.Errorf("comparison operator %s not supported", comparison.Operator.ToString())
+	}
+	return opcode, nil
+}
+
+// compare returns true after applying the comparison specified in the Filter to the actual data in the column
+func compare(comparison Opcode, columnValue, filterValue sqltypes.Value) (bool, error) {
+	// NullsafeCompare returns 0 if values match, -1 if columnValue < filterValue, 1 if columnValue > filterValue
+	result, err := evalengine.NullsafeCompare(columnValue, filterValue)
```
Comment (sougou) [inline go/vt/vttablet/tabletserver/vstreamer/planbuilder.go:149]: Don't we need a compare that handles null values semantics here?

**Comment by rohit-nayak-ps** [comment]
Rebasing against main seems to have messed up this PR. Researcing.

File: go/vt/vttablet/tabletserver/vstreamer/planbuilder.go (hunk 2 above)
Comment (rohit-nayak-ps) [reply go/vt/vttablet/tabletserver/vstreamer/planbuilder.go:154]: Add null value semantics

**Comment by rohit-nayak-ps** [comment]
Recreated PR at https://github.com/vitessio/vitess/pull/8247 since this PR got messed up while rebasing with main.

**Pull request closed without merging**